text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+
+# Alternatives
+[alternatives]: #alternatives
+
+## use `const Trait` bounds for conditionally-const, invent new syntax for always-const
+
+It may seem tempting to use `const fn foo<T: const Trait>` to mean what in this RFC is `~const Trait`, and then add new syntax for bounds that allow using trait methods in const blocks.
+
+Examples of possible always const syntax:
+
+* `=const Trait`
+* `const const Trait` (lol)
+* `const(always) Trait` (`pub` like)
+* `const<true> Trait` (effect generic like)
+* `const! Trait`
+
+## use `Trait<const>` or `Trait<bikeshed#effect: const>` instead of `const Trait`
+
+To avoid new syntax before paths referring to traits, we could treat the constness as a generic parameter or an associated type.
+While an associated type is very close to how the implementation works, neither `effect = const` nor `effect: const` are representing the logic correctly,
+as `const` implies `~const`, but `~const` is nothing concrete, it's more like a generic parameter referring to the constness of the function.
+Fully expanded one can think of
+
+```rust
+const fn foo<T: ~const Trait + const OtherTrait>(t: T) { ... }
+```
+
+to be like
+
+```rust
+const<const C: bool> fn foo<T>(t: T)
+where
+    T: Trait + OtherTrait,
+    <T as Trait>::bikeshed#effect = const<C>,
+    <T as OtherTrait>::bikeshed#effect = const<true>,
+{
+    ...
+}
+```
+
+Note that `const<true>` implies `const<false>` and thus also `for<C> const<C>`, just like `const Trait` implies `~const Trait`.
+
+We do not know of any cases where such an explicit syntax would be useful (only makes sense if you can do math on the bool),
+so a more reduced version could be
+
+```rust
+const fn foo<T>(t: T)
+where
+    T: Trait + OtherTrait,
+    <T as Trait>::bikeshed#effect = ~const,
+    <T as OtherTrait>::bikeshed#effect = const,
+{
+    ...
+}
+```
+
+or
+
+```rust
+const fn foo<T: Trait<bikeshed#effect = ~const> + OtherTrait<bikeshed#effect = const>>(t: T) { ... }
+```
+
+## Make all `const fn` arguments `~const Trait` by default and require an opt out `?const Trait`
+
+We could default to making all `T: Trait` bounds be const if the function is called from a const context, and require a `T: ?const Trait` opt out
+for when a trait bound is only used for its associated types and consts.
+
+This requires a new `~const fn` syntax (sigils or syntax bikesheddable), as the existing `const fn` already has trait bounds that
+do not require const trait impls even if used in const contexts.
+
+An example from libstd today is [the impl block of Vec::new](https://github.com/rust-lang/rust/blob/1ab85fbd7474e8ce84d5283548f21472860de3e2/library/alloc/src/vec/mod.rs#L406) which has an implicit `A: Allocator` bound from [the type definition](https://github.com/rust-lang/rust/blob/1ab85fbd7474e8ce84d5283548f21472860de3e2/library/alloc/src/vec/mod.rs#L397).
+
+A full example how how things would look then
```

Comment:
The example does not use the `~const fn` syntax that the text above says is needed, so I am confused.